### ISSUE TITLE
Simple program for get password

### ISSUE THREAD

**Issue by akibsayyed** [angr/angr#155]
Dear All

I have created simple password program

`#include <stdio.h>
main() {
	char lhs[10];
	char rhs[10]="hithis";
	printf ("\nenter char");
	scanf("%s",&lhs);
	if(strcmp(lhs, rhs)==0)
		printf("correct pass\n");
	else 
		printf("wrong pass\n");
}
`
trying solve it using angr

angr finds solution of i use like this
if(strcmp(lhs, "hithis")==0)

but when i use 
if(strcmp(lhs, rhs)==0)
it gives blank
what am doing wrong

here is my angr script 

`#!/usr/bin/env python2

import angr

AVOID_ADDR = 0x0804858F 
FIND_ADDR = 0x08048581 

def main():
	proj = angr.Project('a.out', load_options={"auto_load_libs": False}) 
	path_group = proj.factory.path_group()
	path_group.explore(find=FIND_ADDR, avoid=AVOID_ADDR) 
	return path_group.found[0].state.posix.dumps(0).split('\0')[0] # stdin
	#return path_group.found[0].state.posix.dumps(0).lstrip('+0').rstrip('B')

def test():
	assert main() == ''

if __name__ == '__main__':
	print(main())`

Please help me

**Comment by schieb** [comment]
tl;dr Your program exhibits undefined behavior and the solution angr gives is correct :smiley:.

The problem you are having stems from the fact that the user input stored in lhs is capable of overwriting the stack allocated string that you have stored in rhs. In fact, it is capable of overwriting much more than just your variables!

angr does not know how many bytes the user will input during the call to scanf, so it uses a default hardcoded max limit (much more than 10 bytes). In addition, angr does not know what those bytes will be. So when the scanf procedure is called, angr is setting not only the 10 bytes of your lhs buffer, but also the next (much - 10) bytes to symbolic values (read: I don't know what these bytes are, but they depend on ```stdin``` in this case). The hard coded string, rhs, that you want to compare to lhs happens to reside in memory right after rhs, so it is now symbolic.

When strcmp is called, instead of comparing the user's string to your hardcoded string, you are comparing two user defined strings. So, in fact, the solution you see is a correct solution.

Try for yourself: ```echo -ne "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" | ./a.out```
(depending on your compiler, you may need to add more null bytes)
Add even more bytes to overwrite your return address for a segfault or program abortion.


Now, in the version where you use ```if(strcmp(lhs, "hithis")==0)```, the compiler stored the string "hithis" in the readonly data section, and a pointer to that string resides on your stack. You are passing that pointer during your call to ```strcmp```.

Just as before, angr does not know how many bytes scanf will read, and results in storing symbolic values over the "hithis" pointer on your stack. Symbolic pointers are difficult to reason about, and in this case, the pointer could literally point anywhere after being overwritten by user data. angr must be intelligent about this and it looks like it chose to ignore any symbolic values that the pointer could be and instead resorted to considering the states in which the pointer is concrete (i.e., program executions in which the user input does not overwrite your pointer).

So how can you get your example to work with a string allocated on the stack?
You will need to add some code to let angr know more information about the user input. Try adding a line to check the string length of the user input and abort execution if the result is not the expected value.

**Comment by zardus** [comment]
Thank you for tracking this down, schieb! Closing this issue :-)
